diff -Nru apache_1.3.12/src/Configuration.tmpl apache_1.3.12.new/src/Configuration.tmpl
--- apache_1.3.12/src/Configuration.tmpl	Tue Mar 14 02:04:30 2000
+++ apache_1.3.12.new/src/Configuration.tmpl	Tue Mar 14 02:05:17 2000
@@ -83,7 +83,7 @@
 #   course.
 
 Rule EAPI=no
-#EAPI_MM=SYSTEM
+EAPI_MM=SYSTEM
 
 ################################################################
 # Dynamic Shared Object (DSO) support
diff -Nru apache_1.3.12/src/include/httpd.h apache_1.3.12.new/src/include/httpd.h
--- apache_1.3.12/src/include/httpd.h	Tue Mar 14 02:04:30 2000
+++ apache_1.3.12.new/src/include/httpd.h	Tue Mar 14 02:07:32 2000
@@ -376,7 +376,7 @@
  */
 #ifdef EAPI
 #ifndef EAPI_MM_CORE_PATH
-#define EAPI_MM_CORE_PATH "logs/mm"
+#define EAPI_MM_CORE_PATH "/var/lib/apache/mm"
 #endif
 #ifndef EAPI_MM_CORE_MAXSIZE
 #define EAPI_MM_CORE_MAXSIZE 1024*1024*1 /* max. 1MB */
