--- httpd-2.0.50/support/apxs.in.orig	2004-07-25 21:25:37.955097000 +0000
+++ httpd-2.0.50/support/apxs.in	2004-07-25 22:25:01.093725664 +0000
@@ -198,11 +198,10 @@
     ($httpd = $0) =~ s:support/apxs$::;
 }
 
-unless (-x "$httpd") {
-	error("$httpd not found or not executable");
-	exit 1;
+if (not -x "$httpd") {
+	print STDERR "Note: $httpd not found or not executable.\n";
 }
-
+else {
 unless (grep /mod_so/, `. $envvars && $httpd -l`) {
     error("Sorry, no shared object support for Apache");
     error("available under your platform. Make sure");
@@ -210,7 +209,7 @@
     error("your server binary `$httpd'.");
     exit 1;
 }
-
+}
 sub get_config_vars{
     my ($file, $rh_config) = @_;
 
