--- httpd-2.0.46/srclib/apr/configure.in.orig	Sat May  3 06:11:07 2003
+++ httpd-2.0.46/srclib/apr/configure.in	Mon Jun 30 13:28:19 2003
@@ -441,10 +441,9 @@
 dnl It should check for LIBS being empty and set LIBS equal to the new value 
 dnl without the extra " " in that case, but they didn't do that.  So, we
 dnl end up LIBS="-lm -lcrypt -lnsl  -ldl" which is an annoyance.
-AC_CHECK_LIB(resolv, res_init)
-AC_CHECK_LIB(nsl, gethostbyname)
+dnl AC_CHECK_LIB(resolv, res_init) -- not used
 AC_SEARCH_LIBS(gethostname, nsl)
-AC_CHECK_LIB(socket, socket)
+AC_SEARCH_LIBS(socket, socket)
 AC_SEARCH_LIBS(crypt, crypt ufc)
 AC_CHECK_LIB(truerand, main)
 AC_SEARCH_LIBS(modf, m)
