--- httpd-2.0.43/Makefile.in.wiget	Fri Oct  4 15:20:09 2002
+++ httpd-2.0.43/Makefile.in	Fri Oct  4 15:20:32 2002
@@ -279,7 +279,7 @@ install-suexec-binary:
 
 install-suexec-setuid:
 	@if test -f $(builddir)/support/suexec; then \
-	    chmod 4755 $(DESTDIR)$(sbindir)/suexec; \
+	    chmod 0755 $(DESTDIR)$(sbindir)/suexec; \
 	fi
 
 install-suexec-caps:

--- httpd-2.4.34/Makefile.in~	2018-07-16 18:45:42.000000000 +0200
+++ httpd-2.4.34/Makefile.in	2018-07-16 19:33:00.579273405 +0200
@@ -277,12 +277,12 @@ install-suexec-binary:
             $(INSTALL_PROGRAM) $(top_builddir)/support/suexec $(DESTDIR)$(sbindir); \
 	fi
 
-install-suexec-setuid:
+install-suexec-setuid: install-suexec-binary
 	@if test -f $(builddir)/support/suexec; then \
 	    chmod 0755 $(DESTDIR)$(sbindir)/suexec; \
 	fi
 
-install-suexec-caps:
+install-suexec-caps: install-suexec-binary
 	@if test -f $(builddir)/support/suexec; then \
             setcap 'cap_setuid,cap_setgid+pe' $(DESTDIR)$(sbindir)/suexec; \
 	fi
