diff -urN httpd-2.2.3.org/server/log.c httpd-2.2.3/server/log.c
--- httpd-2.2.3.org/server/log.c	2006-07-12 05:38:44.000000000 +0200
+++ httpd-2.2.3/server/log.c	2006-09-11 12:12:55.193571822 +0200
@@ -237,8 +237,6 @@
     apr_proc_t *procnew;
 
     if (((rc = apr_procattr_create(&procattr, p)) == APR_SUCCESS)
-        && ((rc = apr_procattr_cmdtype_set(procattr,
-                                           APR_SHELLCMD_ENV)) == APR_SUCCESS)
         && ((rc = apr_procattr_io_set(procattr,
                                       APR_FULL_BLOCK,
                                       APR_NO_PIPE,
@@ -793,8 +791,6 @@
     apr_status_t status;
 
     if (((status = apr_procattr_create(&procattr, pl->p)) != APR_SUCCESS) ||
-        ((status = apr_procattr_cmdtype_set(procattr,
-                                            APR_SHELLCMD_ENV)) != APR_SUCCESS) ||
         ((status = apr_procattr_child_in_set(procattr,
                                              ap_piped_log_read_fd(pl),
                                              ap_piped_log_write_fd(pl)))
