diff -urNp -x '*.orig' httpd-2.4.46.org/configure.in httpd-2.4.46/configure.in
--- httpd-2.4.46.org/configure.in	2021-06-16 09:19:52.228364279 +0200
+++ httpd-2.4.46/configure.in	2021-06-16 09:19:53.851781461 +0200
@@ -873,9 +873,9 @@ echo $MODLIST | $AWK -f $srcdir/build/bu
 APR_EXPAND_VAR(ap_prefix, $prefix)
 AC_DEFINE_UNQUOTED(HTTPD_ROOT, "${ap_prefix}",
 	[Root directory of the Apache install area])
-AC_DEFINE_UNQUOTED(SERVER_CONFIG_FILE, "${rel_sysconfdir}/${progname}.conf",
+AC_DEFINE_UNQUOTED(SERVER_CONFIG_FILE, "${exp_sysconfdir}/apache.conf",
 	[Location of the config file, relative to the Apache root directory])
-AC_DEFINE_UNQUOTED(AP_TYPES_CONFIG_FILE, "${rel_sysconfdir}/mime.types",
+AC_DEFINE_UNQUOTED(AP_TYPES_CONFIG_FILE, "${exp_sysconfdir}/mime.types",
 	[Location of the MIME types config file, relative to the Apache root directory])
 
 perlbin=`$ac_aux_dir/PrintPath perl`
