
Open log files read-only; Jeff Trawick, r1.135 from CVS.

--- httpd-2.0.46/server/log.c.rolog
+++ httpd-2.0.46/server/log.c
@@ -200,7 +200,7 @@
         return APR_EBADPATH;
     }
     if ((rc = apr_file_open(&stderr_file, filename,
-                            APR_APPEND | APR_READ | APR_WRITE | APR_CREATE,
+                            APR_APPEND | APR_WRITE | APR_CREATE,
                             APR_OS_DEFAULT, p)) != APR_SUCCESS) {
         ap_log_error(APLOG_MARK, APLOG_STARTUP, rc, NULL,
                      "%s: could not open error log file %s.",
@@ -313,7 +313,7 @@
             return DONE;
         }
         if ((rc = apr_file_open(&s->error_log, fname,
-                               APR_APPEND | APR_READ | APR_WRITE | APR_CREATE,
+                               APR_APPEND | APR_WRITE | APR_CREATE,
                                APR_OS_DEFAULT, p)) != APR_SUCCESS) {
             ap_log_error(APLOG_MARK, APLOG_STARTUP, rc, NULL,
                          "%s: could not open error log file %s.",
